### PR TITLE
feat: add regional and edge support

### PR DESCRIPTION
**Pull request by eshanholtz**
https://issues.corp.twilio.com/browse/DI-646

### CODE REVIEW

File: test/services/twilio-api/twilio-client.test.js [hunk 1]
```diff
@@ -265,6 +265,200 @@ describe('services', () => {
           expect(options.PageSize).to.be.undefined;
         });
       });
+
+      describe('regional and edge support', () => {
+        let regionTest = test
```
Comment (childish-sambino) [inline test/services/twilio-api/twilio-client.test.js:270]: Don't like reassigning this between tests. IDEs may run this tests in different order and I _think_ this could cause a breakage.

File: src/services/open-api-client.js [hunk 2]
```diff
@@ -44,13 +44,10 @@ class OpenApiClient {
...
-
-        // From 'https://api.twilio.com/' to 'https://api.{region}.twilio.com/'
-        if (parts.length > 1 && parts[1] !== opts.region) {
-          parts.splice(1, 0, opts.region);
-          opts.host = parts.join('.');
-        }
+        const domain = opts.host.split('.').slice(-2).join('.');
+        const prefix = opts.host.split('.' + domain)[0];
+        let product = prefix.split('.')[0];
+        opts.host = [product, opts.edge, opts.region, domain].filter(part => part).join('.');
```
Comment (childish-sambino) [inline src/services/open-api-client.js:50]: How is the original edge being preserved if `opts.edge` is empty?

File: src/services/twilio-api/twilio-client.js [hunk 3]
```diff
@@ -186,6 +187,11 @@ class TwilioApiClient {
       }
     }
 
+    opts.edge = this.edge || opts.edge;
```
[the remark this answers is not in the document]
Comment (eshanholtz) [reply src/services/twilio-api/twilio-client.js:190]: that's opposite of what we've been saying. If the client has a region/edge set, use that. Then use whatever's passed into the request opts

File: src/services/open-api-client.js (hunk 2 above)
Comment (eshanholtz) [reply src/services/open-api-client.js:50]: That logic is [here](https://github.com/twilio/twilio-cli-core/pull/86/files#diff-acfdff5a80e1bd752425fa37d497eb1aR190). `opts.edge = this.edge || opts.edge` will have it take on the client's set value first, followed by whatever's passed into opts. If `opts.edge` is still undefined by this point, it's because it's supposed to be.

File: src/services/twilio-api/twilio-client.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (childish-sambino) [reply src/services/twilio-api/twilio-client.js:190]: We've been talking about different things. Let's sync offline.

File: src/services/config.js [hunk 4]
```diff
@@ -15,14 +15,21 @@ class ConfigDataProfile {
 
 class ConfigData {
   constructor() {
+    this.edge = process.env.TWILIO_EDGE;
```
Comment (childish-sambino) [inline src/services/config.js:18]: I think this will stick the current value of the env var in their config file if it gets saved. Better to default to `null` or `undefined`.

File: src/services/open-api-client.js [hunk 5]
```diff
@@ -97,6 +96,25 @@ class OpenApiClient {
...
+  getHost(host, opts) {
+    if (opts.region || opts.edge) {
+      const domain = host.split('.').slice(-2).join('.');
+      const prefix = host.split('.' + domain)[0];
+      let [product, edge, region] = prefix.split('.');
+      if (edge && !region) {
+        region = edge;
+        edge = undefined;
+      }
+      opts.edge = opts.edge || edge;
```
Comment (childish-sambino) [inline src/services/open-api-client.js:108]: Better not to fiddle with the `opts` if it's easily avoidable. Can just save it to `edge`.

File: src/services/open-api-client.js [hunk 6]
```diff
@@ -97,6 +96,25 @@ class OpenApiClient {
...
+      opts.region = opts.region || region || (opts.edge ? 'us1' : undefined);
```
Comment (childish-sambino) [inline src/services/open-api-client.js:109]: nit: This works too: `opts.region = opts.region || region || (opts.edge && 'us1');`

File: src/services/open-api-client.js [hunk 7]
```diff
@@ -34,26 +35,21 @@ class OpenApiClient {
...
-        }
-      }
-
+      opts.host = this.getHost(opts.host, opts);
       opts.uri = opts.host + opts.uri;
+    } else if (opts.uri) {
+      let uri = new url.URL(opts.uri);
+      uri.hostname = this.getHost(uri.hostname, opts);
+      uri.pathname = this.getUri(uri.pathname, opts);
+      opts.uri = uri.href;
```
Comment (childish-sambino) [inline src/services/open-api-client.js:52]: I think this logic can all be pulled out of the block. If we've made it here, `opts.uri` should be fully formed and we can now use the generic logic of parsing it and updating/inserting the edge and region.

File: src/services/twilio-api/twilio-client.js [hunk 8]
```diff
@@ -180,12 +181,12 @@ class TwilioApiClient {
       opts.headers.Accept = 'application/json';
     }
 
-    if (!opts.uri) {
-      if (opts.path.includes(TwilioApiFlags.ACCOUNT_SID) && !doesObjectHaveProperty(opts.pathParams, TwilioApiFlags.ACCOUNT_SID)) {
-        opts.pathParams[TwilioApiFlags.ACCOUNT_SID] = this.accountSid;
-      }
+    if (opts.path.includes(TwilioApiFlags.ACCOUNT_SID) && !doesObjectHaveProperty(opts.pathParams, TwilioApiFlags.ACCOUNT_SID)) {
+      opts.pathParams[TwilioApiFlags.ACCOUNT_SID] = this.accountSid;
```
Comment (childish-sambino) [inline src/services/twilio-api/twilio-client.js:185]: Why was this changed?

File: test/services/config.test.js [hunk 9]
```diff
@@ -76,7 +76,6 @@ describe('services', () => {
         expect(profile.apiKey).to.equal(constants.FAKE_ACCOUNT_SID);
         expect(profile.apiSecret).to.equal(FAKE_AUTH_TOKEN);
       });
-
```
Comment (childish-sambino) [inline test/services/config.test.js:79]: nooooooooooooo

File: src/services/twilio-api/twilio-client.js (hunk 8 above)
Comment (eshanholtz) [reply src/services/twilio-api/twilio-client.js:185]: path params weren't being substituted if there was a uri. For example, in the tests where `opts.uri` was defined and a path representing the endpoint was provided via `opts.path`, it would not append the path to the url.

Comment (childish-sambino) [reply src/services/twilio-api/twilio-client.js:185]: This only applied to the account SID path param.

Comment (eshanholtz) [reply src/services/twilio-api/twilio-client.js:185]: The endpoint used for the tests did have account sid as the only path param, so I guess it wouldn't be an issue for other endpoints. Why wouldn't we want this substitution made if `opts.uri` is provided?

File: src/services/open-api-client.js (hunk 7 above)
Comment (eshanholtz) [reply src/services/open-api-client.js:52]: Nice catch

File: src/services/twilio-api/twilio-client.js (hunk 8 above)
Comment (childish-sambino) [reply src/services/twilio-api/twilio-client.js:185]: I think the reason the check was there originally is because only the `list` operation would ever supply a `uri` (when getting the next page of results). Other operations would not include a `uri` because it would be built downstream. The next page `uri` should not be parameterized and thus would not need any `pathParams` to be set. In fact, `pathParams` is not a key in `opts` in this case (see `list` implementation above) so attempting to update path params would cause an exception and so a check was added for the uri.

Comment (eshanholtz) [reply src/services/twilio-api/twilio-client.js:185]: So I guess the question is, if we never expect the `opts.uri` to contain a region/edge, should we even account for these situations? Should those scenarios even be tested? The only reason I removed that check was to account for situations were `opts.uri` was of the form `https://api.edge.region.twilio.com`

Comment (childish-sambino) [reply src/services/twilio-api/twilio-client.js:185]: `opts.uri` may or may not contain edge/region. But it should never be parameterized.

Comment (eshanholtz) [reply src/services/twilio-api/twilio-client.js:185]: Wouldn't this break the `nextPageUri` test, then? (note: it did break because of this change, but was fixed with the additional path check in `getUri`

File: src/services/open-api-client.js [hunk 10]
```diff
@@ -34,28 +35,23 @@ class OpenApiClient {
...
-        const parts = opts.host.split('.');
-
-        // From 'https://api.twilio.com/' to 'https://api.{region}.twilio.com/'
-        if (parts.length > 1 && parts[1] !== opts.region) {
-          parts.splice(1, 0, opts.region);
-          opts.host = parts.join('.');
-        }
-      }
-
+      opts.host = this.getHost(opts.host, opts);
```
Comment (childish-sambino) [inline src/services/open-api-client.js:46]: This is being done below.

File: src/services/twilio-api/twilio-client.js (hunk 8 above)
Comment (childish-sambino) [reply src/services/twilio-api/twilio-client.js:185]: Not sure which test that is, but `pathParams` will be undefined in the subsequent calls to `request` from `list`. There are no path params when paging because the `uri` is expected to be fully-formed already.

File: src/services/config.js [hunk 11]
```diff
@@ -130,6 +139,7 @@ class ConfigData {
   }
 
   loadFromObject(configObj) {
+    this.edge = this.edge || configObj.edge;
```
[the remark this answers is not in the document]
Comment (eshanholtz) [reply src/services/config.js:142]: When we talked about order of precedence we decided 1. opts.edge, 2. env vars, 3. config, 4. URI edge
Are you saying precedence should be changed? What about region?

[the remark this answers is not in the document]
Comment (childish-sambino) [reply src/services/config.js:142]: `loadFromObject` is taking what's in the `config.json` and populating the `ConfigData` object. When loading config, it shouldn't care what's in the object already; the config file contains everything needed to build the config object.

File: src/services/twilio-api/twilio-client.js (hunk 8 above)
Comment (eshanholtz) [reply src/services/twilio-api/twilio-client.js:185]: `opts.pathParams` defaults to `opts.pathParams || {}`. Wouldn't that mean in subsequent calls, `opts.pathParams === {}`, which would just fail the condition in the if?

Comment (childish-sambino) [reply src/services/twilio-api/twilio-client.js:185]: Ah, I missed that line. But either way, the original intent was to illustrate that when a `uri` is given, `pathParams` don't do anything.

Comment (eshanholtz) [reply src/services/twilio-api/twilio-client.js:185]: Ah, understood 👍

File: src/services/config.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (eshanholtz) [reply src/services/config.js:142]: Ok that makes sense

File: src/services/open-api-client.js (hunk 10 above)
Comment (eshanholtz) [reply src/services/open-api-client.js:46]: The one below doesn't take `opts.host` into account, it just parses from `opts.uri`

Comment (childish-sambino) [reply src/services/open-api-client.js:46]: Yeah, but the line after this one uses `opts.host`.

Comment (eshanholtz) [reply src/services/open-api-client.js:46]: 🤦 I see it now

File: src/base-commands/twilio-client-command.js [hunk 12]
```diff
@@ -111,7 +111,8 @@ class TwilioClientCommand extends BaseCommand {
   buildClient(ClientClass) {
     return new ClientClass(this.currentProfile.apiKey, this.currentProfile.apiSecret, {
       accountSid: this.flags[CliFlags.ACCOUNT_SID] || this.currentProfile.accountSid,
-      region: this.currentProfile.region,
+      edge: process.env.TWILIO_EDGE || this.userConfig.edge,
+      region: process.env.TWILIO_REGION || this.currentProfile.region,
```
Comment (childish-sambino) [inline src/base-commands/twilio-client-command.js:115]: I don't think we should pull the region env var here. Since region is tied to auth, the profile object should contain the correct region already.

File: src/services/config.js [hunk 13]
```diff
@@ -156,13 +166,13 @@ class Config {
     if (!fs.existsSync(this.filePath)) {
       return configData;
     }
-
```
Comment (childish-sambino) [inline src/services/config.js:159]: noooooooooooo

File: src/services/config.js [hunk 14]
```diff
@@ -156,13 +166,13 @@ class Config {
...
     configData.loadFromObject(await fs.readJSON(this.filePath));
     return configData;
   }
 
   async save(configData) {
     configData = {
+      edge: configData.edge,
       email: configData.email,
       prompts: configData.prompts,
       // Note the historical 'projects' naming.
```
Comment (childish-sambino) [inline src/services/config.js:178]: Should go ahead and fix the migration comment below.

File: src/services/open-api-client.js [hunk 15]
```diff
@@ -34,28 +35,22 @@ class OpenApiClient {
...
-          opts.host = parts.join('.');
-        }
-      }
-
       opts.uri = opts.host + opts.uri;
     }
 
+    const uri = new url.URL(opts.uri);
+    uri.hostname = this.getHost(uri.hostname, opts);
+    uri.pathname = this.getUri(uri.pathname, opts);
```
Comment (childish-sambino) [inline src/services/open-api-client.js:51]: I don't think this line is needed (along with the changes to `getUri`). The path should already be fully-formed at this point.

File: test/base-commands/twilio-client-command.test.js [hunk 16]
```diff
@@ -230,5 +234,69 @@ describe('base-commands', () => {
...
+        return test
+          .do(ctx => {
+            ctx.userConfig = new ConfigData();
+            if (configEdge) {
+              ctx.userConfig.edge = configEdge;
+            }
+            if (envRegion) {
+              process.env.TWILIO_REGION = envRegion;
+            }
+            if (envEdge) {
```
Comment (childish-sambino) [inline test/base-commands/twilio-client-command.test.js:252]: These `if` checks aren't necessary.

File: test/base-commands/twilio-client-command.test.js [hunk 17]
```diff
@@ -230,5 +234,69 @@ describe('base-commands', () => {
...
+              process.env.TWILIO_EDGE = envEdge;
+            }
+            ctx.userConfig.addProfile('default-profile', constants.FAKE_ACCOUNT_SID);
+            ctx.userConfig.addProfile('region-edge-testing', constants.FAKE_ACCOUNT_SID, 'configRegion');
```
Comment (childish-sambino) [inline test/base-commands/twilio-client-command.test.js:256]: `configRegion` is a string literal but `configEdge` is a param?

File: src/base-commands/twilio-client-command.js (hunk 12 above)
Comment (eshanholtz) [reply src/base-commands/twilio-client-command.js:115]: Ah, forgot it was pulled when it pulls the other profile info from env vars.

File: test/services/twilio-api/twilio-client.test.js [hunk 18]
```diff
@@ -265,6 +265,230 @@ describe('services', () => {
...
+
+          const response = await client.create({
+            domain: 'api',
+            path: '/2010-04-01/Accounts/{AccountSid}/Messages.json',
+            region: 'region',
+            // Should ignore the region in the uri
+            uri: 'https://api.uriRegion.twilio.com'
+          });
+          expect(response).to.eql({ status: 'queued' });
+        });
```
Comment (childish-sambino) [inline test/services/twilio-api/twilio-client.test.js:357]: Looks just like the previous test.

File: test/services/twilio-api/twilio-client.test.js [hunk 19]
```diff
@@ -265,6 +265,230 @@ describe('services', () => {
...
+            constants.FAKE_API_KEY, constants.FAKE_API_SECRET,
+            { accountSid, httpClient }
+          );
+
+          const response = await client.create({
+            domain: 'api',
+            path: '/2010-04-01/Accounts/{AccountSid}/Messages.json',
+            uri: 'https://api.edge.region.twilio.com'
+          });
+          expect(response).to.eql({ status: 'queued' });
```
Comment (childish-sambino) [inline test/services/twilio-api/twilio-client.test.js:489]: Nice test coverage!!!

File: test/base-commands/twilio-client-command.test.js (hunk 16 above)
Comment (eshanholtz) [reply test/base-commands/twilio-client-command.test.js:252]: Yeah they are. If I don't check, it sets process.env.TWILIO_WHATEVER to the string `undefined`.

Comment (childish-sambino) [reply test/base-commands/twilio-client-command.test.js:252]: But that's fine, right?

Comment (childish-sambino) [reply test/base-commands/twilio-client-command.test.js:252]: Well, why would it be a string?

File: test/base-commands/twilio-client-command.test.js (hunk 17 above)
Comment (eshanholtz) [reply test/base-commands/twilio-client-command.test.js:256]: Well, I wanted it to be tied to that profile regardless of whether that profile's active. I'll make it a param and gave it a default value of `'configRegion'`

File: test/services/twilio-api/twilio-client.test.js (hunk 18 above)
Comment (eshanholtz) [reply test/services/twilio-api/twilio-client.test.js:357]: The previous test has the `client.region`, `opts.region` and uri region all defined, and asserts that `opts.region` takes precedence. This test does not have `client.region` defined and asserts it uses `opts.region`

File: src/services/open-api-client.js (hunk 15 above)
Comment (eshanholtz) [reply src/services/open-api-client.js:51]: Hmm.. when I take it out it breaks the tests..

File: test/services/twilio-api/twilio-client.test.js (hunk 19 above)
Comment (eshanholtz) [reply test/services/twilio-api/twilio-client.test.js:489]: thanks!

File: src/services/config.js (hunk 14 above)
Comment (eshanholtz) [reply src/services/config.js:178]: Not sure I understand what this means

File: src/services/config.js (hunk 13 above)
Comment (eshanholtz) [reply src/services/config.js:159]: omg it's not even a _good_ use of whitespace

File: src/services/config.js (hunk 14 above)
Comment (childish-sambino) [reply src/services/config.js:178]: There's a line further down about Node8 (which is no longer supported).

File: src/services/open-api-client.js (hunk 15 above)
Comment (childish-sambino) [reply src/services/open-api-client.js:51]: Is there a test passing a parameterized `opts.uri`?

File: test/base-commands/twilio-client-command.test.js (hunk 16 above)
Comment (eshanholtz) [reply test/base-commands/twilio-client-command.test.js:252]: I'm not sure. I haven't been able to figure that one out

File: src/services/open-api-client.js (hunk 15 above)
Comment (eshanholtz) [reply src/services/open-api-client.js:51]: Not a parameterized `opts.uri`, but we have a parameterized `opts.path`

Comment (childish-sambino) [reply src/services/open-api-client.js:51]: Line 38 should eval the `opts.path` there.
